source/extensions/filters/http/ratelimit/ratelimit.cc

```diff
@@ -194,9 +195,16 @@ void Filter::complete(Filters::Common::RateLimit::LimitStatus status,
   if (status == Filters::Common::RateLimit::LimitStatus::OverLimit &&
       config_->runtime().snapshot().featureEnabled("ratelimit.http_filter_enforcing", 100)) {
     state_ = State::Responded;
+    // Always overwrite the content-type automatically set by sendLocalReply whenever
+    // we're sending back a response body here. We do this because any content-type
+    // coming from the ratelimit service should be treated as authoritative, and we
+    // must discard the default text/plain type set by default.
+    const bool overwrite_content_type = response_body.length() > 0;
```

Comment:
What's the reasoning for not treating a content-type in the RateLimiterResponse as authoritative when the body is empty?

Reply:
The ratelimit service can only specify response headers to add, without the option to append or override. Because there's no option, the behavior is "append". But append doesn't really make sense for content-type here when we know sendLocalReply is going to always add with text/plain. Then, later, our code may add the actual content-type as it was received from the RLS.

The response header config is here https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ratelimit/v3/rls.proto#service-ratelimit-v3-ratelimitresponse - note HeaderValue instead of HeaderValueOption

I think it makes sense to make the content-type overwrite any value present here. Maybe the right thing to do long term is both to migrate the RLS service to use HeaderValueOptions (just because, so it is similar to ext_authz) and also to clean up the interaction between sendLocalReply and this function.

If we rename `overwrite_content_type` to `is_local_response` (or just `local_response`), do you think the behavior can/should stay as it is implemented here?

Reply:
Sorry for the slow reply, I got slammed this week.

I think you're right that the RLS should be able to specified whether headers should be replaced or appended.

I'm just trying to figure out if there's a use case for a rejection setting content-type with an empty body. I think I would find it surprising if it didn't overwrite just because there was no body. And thus, I think if you do the param rename, for the local response case you should treat content-type from RLS as authoritative even if body is missing.

Reply:
Sounds good. I will rename the variable to `from_local_reply` (or something else if you prefer) and then go ahead and make the content-type authoritative in that case, even with no body.

Do you think there's value in following up with a PR to migrate the RLS to HeaderValueOptions? If so, what's your opinion on migration/strategy? I could either add a new field or possibly make a breaking change for v4, right?

Reply:
I don't feel I'm in a position to pass judgement on whether switching to HeaderValueOption is worth the migration effort for users. To the extent that you can find people using RLS and poll them, I think that'd be the way to go.

As far as the mechanics of the change go, I think breaking changes are only allowed within an alpha version if the field was never part of a released version. So you wouldn't be allowed to do that here. I think the way we'd do this is to add the new field and mark the existing one deprecated. When v4 becomes authoritative, we'd add translation to convert the deprecated v3 field under the covers with warnings, stats, etc (there's precedent for all this).

Reply:
That all makes sense. Thanks for the explanation. I will look into adding a new field.